### PR TITLE
some improvement

### PR DESCRIPTION
**Pull request by qinzuoyan**

### CODE REVIEW

File: src/core/core/network.cpp [hunk 1]
```diff
@@ -240,13 +255,15 @@ namespace dsn
...
-        _message_count.fetch_add(1, std::memory_order_relaxed); // -- in unlink_message
 
-        msg->add_ref(); // released in on_send_completed        
+        msg->add_ref(); // released in on_send_completed
+
         uint64_t sig;
         {
             utils::auto_lock<utils::ex_lock_nr> l(_lock);
             msg->dl.insert_before(&_messages);
+            _message_count.fetch_add(1, std::memory_order_relaxed); // -- in unlink_message
```
Comment (imzhenyu) [inline src/core/core/network.cpp]: putting this out of the lock to reduce locking size

File: src/core/core/rpc_engine.cpp [hunk 2]
```diff
@@ -133,8 +133,7 @@ namespace dsn {
         if (nullptr == reply)
         {
             call->set_delay(delay_ms);
-            // TODO(qinzuoyan): maybe set err as ERR_NETWORK_FAILURE to differ with ERR_TIMEOUT
-            call->enqueue(ERR_TIMEOUT, reply);
+            call->enqueue(ERR_NETWORK_FAILURE, reply);
```
Comment (imzhenyu) [inline src/core/core/rpc_engine.cpp:148]: not sure our replication libs handle this appropriately now

File: src/core/core/network.cpp [hunk 3]
```diff
@@ -352,27 +369,28 @@ namespace dsn
...
-        
         else
         {
-            rpc_session_ptr sp = this;
-            _net.on_server_session_disconnected(sp);
-        }
-
-        if (is_write)
-        {
-            clear_send_queue(false);
```
Comment (imzhenyu) [inline src/core/core/network.cpp:372]: need more time to check whether removing this is ok.

File: src/core/core/rpc_engine.cpp (hunk 2 above)
Comment (qinzuoyan) [inline src/core/core/rpc_engine.cpp:148]: I have checked again, I think there is no problem now.

File: src/core/core/network.cpp (hunk 1 above)
Comment (qinzuoyan) [inline src/core/core/network.cpp]: I think it's better to access _message_count all in locks, and not use atomic type.

File: src/core/core/network.cpp (hunk 3 above)
Comment (qinzuoyan) [inline src/core/core/network.cpp:372]: In my changes, the on_disconnected() will be invoked once and only once when connection failed. So, we do not distinguish "is_write" here.

**Comment by qinzuoyan** [comment]
@imzhenyu , some fixes as above.

**Comment by imzhenyu** [comment]
@qinzuoyan Thanks for the udpates. I'll look into them in more details later after this busy week :-(

**Comment by qinzuoyan** [comment]
OK, hope for your good news.

File: src/core/core/network.cpp [hunk 4]
```diff
@@ -344,35 +360,36 @@ namespace dsn
...
+            if (is_client())
+            {
+                _net.on_client_session_disconnected(sp);
+            }
+            else
+            {
+                _net.on_server_session_disconnected(sp);
+            }
+
+            clear_send_queue(false);
```
Comment (imzhenyu) [inline src/core/core/network.cpp:385]: @qinzuoyan this change introduces bugs - considering the case where messages are being sent while the remote peer disconnected which causes the read operation callback calls this function first, which immediately causes invalid data access in write operations (e.g., using _sending_buffers). I'll come up the fix later.

**Pull request merged**